resolver/upstream_resolver.go

```diff
@@ -153,13 +156,26 @@ func (r *dnsUpstreamClient) callExternal(msg *dns.Msg,
 }
 
 // NewUpstreamResolver creates new resolver instance
-func NewUpstreamResolver(upstream config.Upstream) *UpstreamResolver {
-	upstreamClient, upstreamURL := createUpstreamClient(upstream)
+// We pre-resolve the server's IP using bootstrap, and then it is automatically
+// refreshed using itself (this new UpstreamResolver).
+func NewUpstreamResolver(upstream config.Upstream, bootstrap *Bootstrap) (*UpstreamResolver, error) {
```

Comment:
The comment mentions self-refreshing the IP. It's just an idea I had: instead of using the bootstrap DNS every time like before, we could use the upstream itself.

I could also make the code similar to before where it does a lookup on each `Dial` and we rely on the `Bootstrap`'s cache. The reason I didn't is bootstrap would need special code to avoid trying to lookup its IP via itself and going into infinite recursion.
I think having each `UpstreamResolver` self-refresh is the best way to keep that code unified and have periodic IP updates.

Also asking each resolver for its own IP sounds like a good idea to me, but I'm not 100% sure.
The potential downside I can imagine would be users that use a bootstrap DNS they specially configured, but I don't think this sounds likely, and if we update the docs to say that the bootstrap DNS is only used for the initial lookup there'd be no surprise.

Implementation-wise, it'd just be a goroutine per resolver that in a loop:
- waits for the TTL of the upstream
- does a lookup for its hostname
- updates its client if needed

And no goroutine if the upstream was just configured with an IP.

What do you think?

Reply:
The issue with that is avoiding recursion with the bootstrap: its chain contains a `UpstreamResolver` which will call the bootstrap, thus itself.
I didn't try to find a solution to this though, so I'll give it a shot because agree that it'd be nicer to have more of the code outside the `UpstreamResolver`.